### PR TITLE
fix(tui): avoid hard-wrapping URL lines in history

### PR DESCRIPTION
**Pull request by danielsacosta**
## What
- Avoid hard-wrapping URL lines in history in TUI/TUI2.
- Fix wrapped row counts for long lines (soft wrap) in history rendering.
- Add tests for long URL wrapping.

## Why
- Bug report link breaks when terminal width causes wrapping; command-click selects only partial URL.
- Issue: https://github.com/openai/codex/issues/7011

## How
- Skip word wrapping for lines containing URLs in history insertion.
- Use row count based on line width and viewport width.
- Add unit tests in TUI and TUI2.

## Testing
- just fmt
- just fix -p codex-tui
- just fix -p codex-tui2
- cargo test -p codex-tui
- cargo test -p codex-tui2

### CODE REVIEW

**Comment by danielsacosta** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by etraut-openai** [comment]
@codex review

File: codex-rs/tui/src/insert_history.rs [hunk 1]
```diff
@@ -318,6 +360,41 @@ mod tests {
         );
     }
 
+    #[test]
+    fn wrap_history_lines_preserves_long_url() {
```
Comment (joshka-oai) [inline codex-rs/tui/src/insert_history.rs]: I'd add tests here that show how we handle something similar to:
"Here is some text and a http://foobar url in the middle"
This seems like it would be a fairly common use case for when we render URLs in output and something that should be explicitly handled.

File: codex-rs/tui/src/insert_history.rs [hunk 2]
```diff
@@ -22,6 +23,46 @@ use ratatui::style::Modifier;
...
+    text.contains("https://") || text.contains("http://")
+}
+
+fn wrap_history_lines<'a>(lines: &'a [Line<'a>], width: u16) -> Vec<Line<'a>> {
+    let base_opts = RtOptions::new(width.max(1) as usize);
+    let mut wrapped: Vec<Line<'a>> = Vec::new();
+    let mut first = true;
+    for line in lines {
+        if line_contains_url(line) {
+            wrapped.push(line.clone());
```
Comment (joshka-oai) [inline codex-rs/tui/src/insert_history.rs]: I don't think we should just not wrap here. What about instead we detect URLs and ensure that we wrap before / after them?

File: codex-rs/tui/src/insert_history.rs [hunk 3]
```diff
@@ -22,6 +23,46 @@ use ratatui::style::Modifier;
...
+
+fn wrapped_row_count(lines: &[Line<'_>], width: u16) -> u16 {
+    let width = width.max(1) as usize;
+    let mut rows = 0usize;
+    for line in lines {
+        let line_width = line.width().max(1);
+        rows = rows.saturating_add(line_width.div_ceil(width));
+    }
+    u16::try_from(rows).unwrap_or(u16::MAX)
+}
```
Comment (joshka-oai) [inline codex-rs/tui/src/insert_history.rs:55]: Can you explain this a bit (better yet - add doc comments that explain this - what are we doing here and why)

**Review by joshka-oai — CHANGES_REQUESTED** [review]
I think we can go a bit deeper here and make an intelligent wrap before URLs a thing - i.e. when they'd wrap in the middle of the URL, force a wrap before. We will still need to character wrap when they're too long for the terminal regardless (and need tests for that scenario), but this would at least cut down the problems of not quite enough space to wrap. In effect, we want to treat the url as a word / non-breakable token somehow. This should probably also be in the wrapping code, not the insert_history code btw.

Strong support for fixing this. So please continue on this when you can.

**Review by joshka-oai — CHANGES_REQUESTED** [review]
I think we can go a bit deeper here and make an intelligent wrap before URLs a thing - i.e. when they'd wrap in the middle of the URL, force a wrap before. We will still need to character wrap when they're too long for the terminal regardless (and need tests for that scenario), but this would at least cut down the problems of not quite enough space to wrap. In effect, we want to treat the url as a word / non-breakable token somehow. This should probably also be in the wrapping code, not the insert_history code btw.

Strong support for fixing this. So please continue on this when you can.

File: codex-rs/tui/src/insert_history.rs (hunk 1 above)
Comment (danielsacosta) [reply codex-rs/tui/src/insert_history.rs]: Thanks for the suggestion. I added a test that covers the ‘Here is some text and a http://foobar url in the middle’ case and verifies we wrap before the URL when it fits. I also added a case where the URL is longer than the width to ensure character‑wrap. The URL logic is now in the shared wrapping code, and insert_history uses it.

File: codex-rs/tui/src/insert_history.rs (hunk 2 above)
Comment (danielsacosta) [reply codex-rs/tui/src/insert_history.rs]: Agreed — I removed the ‘skip wrapping’ branch and now detect URLs in the shared wrapping code. URLs are treated as a single token so we wrap before/after them when they fit; if they exceed the width we fall back to character‑wrap. insert_history now always uses the URL‑aware wrapping.

File: codex-rs/tui/src/insert_history.rs (hunk 3 above)
Comment (danielsacosta) [reply codex-rs/tui/src/insert_history.rs:55]: Added a doc comment above wrapped_row_count explaining that it computes the number of terminal rows needed after wrapping, and why we still account for long tokens (like URLs) that the terminal may hard‑wrap. This keeps history insertion and scroll calculations accurate.

File: codex-rs/tui/src/insert_history.rs [hunk 4]
```diff
@@ -22,6 +23,37 @@ use ratatui::style::Modifier;
 use ratatui::text::Line;
 use ratatui::text::Span;
 
+fn wrap_history_lines<'a>(lines: &'a [Line<'a>], width: u16) -> Vec<Line<'a>> {
```
Comment (joshka-oai) [inline codex-rs/tui/src/insert_history.rs:26]: needs a doc comment clarifying intent

Comment (joshka-oai) [inline codex-rs/tui/src/insert_history.rs:26]: nit (subjective preference): I'd prefer to see this below where it's called as it makes the code easier generally to read top-to-bottom outside-in.

File: codex-rs/tui/src/insert_history.rs [hunk 5]
```diff
@@ -22,6 +23,37 @@ use ratatui::style::Modifier;
...
+    let mut wrapped: Vec<Line<'a>> = Vec::new();
+    let mut first = true;
+    for line in lines {
+        let opts = if first {
+            base_opts.clone()
+        } else {
+            base_opts
+                .clone()
+                .initial_indent(base_opts.subsequent_indent.clone())
+        };
```
Comment (joshka-oai) [inline codex-rs/tui/src/insert_history.rs:37]: This seems a bit odd. I'm not sure I understand why we need to do this rather than relying on initial indent directly here.

File: codex-rs/tui/src/insert_history.rs [hunk 6]
```diff
@@ -22,6 +23,37 @@ use ratatui::style::Modifier;
...
+}
+
+/// Counts the number of terminal rows needed after wrapping `lines` at `width`.
+/// This accounts for any long tokens (like URLs) that are still forced to
+/// hard-wrap by the terminal even after word-aware wrapping.
+fn wrapped_row_count(lines: &[Line<'_>], width: u16) -> u16 {
+    let width = width.max(1) as usize;
+    let mut rows = 0usize;
+    for line in lines {
+        let line_width = line.width().max(1);
```
Comment (joshka-oai) [inline codex-rs/tui/src/insert_history.rs:51]: I don't love the `width()` call here (it's likely not a significant perf hit, but becomes one easily with large enough history which can cause rendering slowdowns - more-so in tui2 than tui.

Not a big deal, but if there's a simple fix to avoid this (like storing the line + width computed once here) that would be great.

File: codex-rs/tui/src/insert_history.rs [hunk 7]
```diff
@@ -22,6 +23,37 @@ use ratatui::style::Modifier;
...
+        rows = rows.saturating_add(line_width.div_ceil(width));
+    }
```
Comment (joshka-oai) [inline codex-rs/tui/src/insert_history.rs:53]: This can probably be more written more succinctly / obviously with map + sum

File: codex-rs/tui/src/wrapping.rs
Comment (joshka-oai) [inline codex-rs/tui/src/wrapping.rs:1]: Add covering tests (make codex produce these so that all the edge cases / behavior are properly tested).

**Review by joshka-oai — CHANGES_REQUESTED** [review]
When I tried this out in both tui and tui2, I asked codex to produce a url that was long enough to wrap. I saw the wrapping seemed like it still happened mid url. It would probably be good to add some tests that use insta around these things (as these tend to be visual rather than programmatic).

tui
Image

tui2
Image

**Comment by joshka-oai** [comment]
Hey @Daniel-Santiago-Acosta-1013 just checking to see if you have had a chance to think about any of the feedback above. No rush.

**Comment by etraut-openai** [comment]
Closing because of lack of contributor response.
